### PR TITLE
ENH: Make RangeIndex.append() return RangeIndex when possible

### LINKED ISSUE #16212: RangeIndex.append() should return RangeIndex when possible
#### Code Sample, a copy-pastable example if possible

```python
In [2]: i1, i2 = pd.RangeIndex(2), pd.RangeIndex(2, 4)

In [3]: i1.union(i2)
Out[3]: RangeIndex(start=0, stop=4, step=1)

In [4]: i1.append(i2)
Out[4]: Int64Index([0, 1, 2, 3], dtype='int64')
```
#### Problem description

It would be nice if appending consecutive ``RangeIndex`` returned a ``RangeIndex``.

This was briefly discussed [here](https://github.com/pandas-dev/pandas/pull/12289#r71085353).

#### Expected Output

Same as ``Out[3]:``

PR on its way.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: f1549668c84b481a67547bb46261d1394c8dd3e7
python: 3.5.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.7.0-1-amd64
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: it_IT.utf8
LOCALE: it_IT.UTF-8

pandas: 0.20.0rc1+49.gf1549668c
pytest: 3.0.6
pip: 9.0.1
setuptools: 33.1.1
Cython: 0.25.2
numpy: 1.12.0
scipy: 0.18.1
xarray: 0.9.1
IPython: 5.1.0.dev
sphinx: 1.4.9
patsy: 0.3.0-dev
dateutil: 2.5.3
pytz: 2016.7
blosc: None
bottleneck: 1.2.0
tables: 3.3.0
numexpr: 2.6.1
feather: 0.3.1
matplotlib: 2.0.0
openpyxl: 2.3.0
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.6
lxml: 3.7.1
bs4: 4.5.3
html5lib: 0.999999999
sqlalchemy: 1.0.15
pymysql: None
psycopg2: None
jinja2: 2.8
s3fs: None
pandas_gbq: None
pandas_datareader: 0.2.1

</details>

### PR DESCRIPTION
**Pull request by toobaz**
- [x] closes #16212
 - [x] tests added / passed
 - [x] passes ``git diff master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

This is analogous to what is already done for ``RangeIndex.union()``, but there doesn't seem to be much scope for code reuse (what could be, in principle, reused in ``.append`` of all ``Index`` subclasses is the initial casting to list/name handling, I could do in a separate PR).

### CODE REVIEW

File: doc/source/whatsnew/v0.20.0.txt [hunk 1]
```diff
@@ -508,6 +508,7 @@ Other Enhancements
 - Compatability with Jupyter notebook 5.0; MultiIndex column labels are left-aligned and MultiIndex row-labels are top-aligned (:issue:`15379`)
 
 - ``TimedeltaIndex`` now has a custom datetick formatter specifically designed for nanosecond level precision (:issue:`8711`)
+- ``RangeIndex.append`` now returns a ``RangeIndex`` object when possible (:issue:`16212`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: move to 0.20.1

File: pandas/core/indexes/range.py [hunk 2]
```diff
@@ -443,6 +443,63 @@ def join(self, other, how='left', level=None, return_indexers=False,
         return super(RangeIndex, self).join(other, how, level, return_indexers,
                                             sort)
 
+    def append(self, other):
+        """
+        Append a collection of Index options together
+
+        Parameters
```
Comment (jreback) [inline pandas/core/indexes/range.py]: hmm, yeah any way to share some code?

Comment (toobaz) [inline pandas/core/indexes/range.py]: yes, more or less ready, but it involves many more changes/files and probably some discussion, if it's OK I will open another PR immediately after this

Comment (jreback) [inline pandas/core/indexes/range.py]: You can simply put up another PR which bases on this one.

File: doc/source/whatsnew/v0.20.1.txt [hunk 3]
```diff
@@ -18,6 +18,7 @@ Highlights include:
 
 Enhancements
 ~~~~~~~~~~~~
+- ``RangeIndex.append`` now returns a ``RangeIndex`` object when possible (:issue:`16212`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.1.txt]: move to 0.21.0

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (toobaz) [inline pandas/core/indexes/range.py]: I have to fix one error

**Comment by toobaz** [comment]
The last push just makes few lines of the tests nicer

File: pandas/core/indexes/range.py (hunk 2 above)
Comment (toobaz) [inline pandas/core/indexes/range.py]: see #16236

**Comment by toobaz** [comment]
@jreback as I had anticipated, #16236 involves some non-trivial discussion (and additional work which I won't be able to do in the next days). For the moment, I invite you to reconsider this PR, which should be straightforward: after, it will be easier to focus on that pure refactoring PR.

File: pandas/tests/indexes/test_range.py [hunk 4]
```diff
@@ -941,3 +941,38 @@ def test_where_array_like(self):
         for klass in klasses:
             result = i.where(klass(cond))
             tm.assert_index_equal(result, expected)
+
+    def test_append(self):
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/test_range.py]: Can you add the issue number as comment?

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
ok, fine to merge. ping on green.

File: pandas/tests/indexes/test_range.py (hunk 4 above)
Comment (toobaz) [inline pandas/tests/indexes/test_range.py]: (done)

**Comment by toobaz** [comment]
ping

**Comment by jreback** [comment]
thanks!
